js/src/instructions/withdraw.ts

```diff
@@ -0,0 +1,102 @@
+import * as BufferLayout from '@solana/buffer-layout';
+import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
+import {
+  PublicKey,
+  StakeProgram,
+  SystemProgram,
+  SYSVAR_CLOCK_PUBKEY,
+  TransactionInstruction,
+} from '@solana/web3.js';
+import BN from 'bn.js';
+import type { Lamports, ProgramAddresses, Snapshot } from '../types';
+import {
+  findAuthorityProgramAddress,
+  getHeaviestValidatorStakeAccount,
+} from '../utils';
+
+export const getWithdrawInstruction = async (
+  snapshot: Snapshot,
+  programAddresses: ProgramAddresses,
+  stSolAccountOwnerAddress: PublicKey,
+  senderStSolAccountAddress: PublicKey,
+  recipientStakeAccountAddress: PublicKey,
+  amount: Lamports
+) => {
+  const stakeAuthorityAddress = await findAuthorityProgramAddress(
+    programAddresses,
+    'stake_authority'
+  );
+  const heaviestValidatorStakeAccount =
+    getHeaviestValidatorStakeAccount(snapshot);
+
+  if (amount.lamports.lte(snapshot.stakeAccountRentExemptionBalance.lamports)) {
+    throw new Error('Amount must be greater than the rent exemption balance');
+  }
+
+  const maxWithdrawAmount = heaviestValidatorStakeAccount.balance.lamports
+    .div(new BN(10))
+    .add(new BN(10));
+
+  if (amount.lamports.gte(maxWithdrawAmount)) {
```

Comment:
This is mixing things in different units (and that is where the different wrapper types come in, to prevent this). Is it possible to overload the `<=` operator in TypeScript? Then we can make sure that for `Lamports` on the left-hand side, the right-hand side must also be `Lamports`, and the compiler will reject `amount < maxWithdrawAmount` with a type error, because `amount: StLamports` but `maxWithdrawAmount: Lamports`. This way we can leverage the type system to prevent bugs :-)

To fix the bug, we need to convert the stSOL amount to SOL, according to the exchange rate in the snapshot.

Reply:
Ahhh yes, that make sense. Will try to fix it